spiffe/tls_peer.go

```diff
@@ -0,0 +1,319 @@
+}
+
+func (w *tlsPeerWatcher) OnError(err error) {
+	w.p.onError(err)
+}
+
+func (p *TLSPeer) Dial(ctx context.Context, network, address string, expectPeer ExpectPeerFunc) (net.Conn, error) {
+	conn, err := new(net.Dialer).DialContext(ctx, network, address)
+	if err != nil {
+		return nil, err
+	}
+
+	config, err := p.GetConfig(ctx, expectPeer)
+	if err != nil {
+		conn.Close()
+		return nil, err
+	}
+
+	return tls.Client(conn, config), nil
+}
+
+func (p *TLSPeer) Listen(ctx context.Context, network, address string, expectPeer ExpectPeerFunc) (net.Listener, error) {
+	inner, err := net.Listen(network, address)
+	if err != nil {
+		return nil, err
+	}
+
+	l, err := p.NewListener(ctx, inner, expectPeer)
+	if err != nil {
+		inner.Close()
+		return nil, err
+	}
+
+	return l, nil
+}
+
+func (p *TLSPeer) NewListener(ctx context.Context, inner net.Listener, expectPeer ExpectPeerFunc) (net.Listener, error) {
+	config, err := p.GetConfig(ctx, expectPeer)
+	if err != nil {
+		return nil, err
+	}
+
+	return &peerTLSListener{
+		Listener: inner,
+		config:   config,
+	}, nil
+}
+
+func (p *TLSPeer) GetConfig(ctx context.Context, expectPeer ExpectPeerFunc) (*tls.Config, error) {
+	if expectPeer == nil {
+		return nil, errors.New("authorize callback is required")
+	}
+	if err := p.WaitUntilReady(ctx); err != nil {
+		return nil, err
+	}
+	return &tls.Config{
+		ClientAuth:            tls.RequireAnyClientCert,
+		InsecureSkipVerify:    true,
+		GetCertificate:        AdaptGetCertificate(p),
+		GetClientCertificate:  AdaptGetClientCertificate(p),
+		VerifyPeerCertificate: AdaptVerifyPeerCertificate(p, expectPeer),
+	}, nil
+}
+
+func AdaptGetCertificate(p *TLSPeer) func(*tls.ClientHelloInfo) (*tls.Certificate, error) {
+	return func(*tls.ClientHelloInfo) (*tls.Certificate, error) {
+		return p.GetCertificate()
+	}
+}
+
+func AdaptGetClientCertificate(p *TLSPeer) func(*tls.CertificateRequestInfo) (*tls.Certificate, error) {
+	return func(*tls.CertificateRequestInfo) (*tls.Certificate, error) {
+		return p.GetCertificate()
+	}
+}
+
+func AdaptVerifyPeerCertificate(p *TLSPeer, expectPeer ExpectPeerFunc) func([][]byte, [][]*x509.Certificate) error {
+	return func(rawCerts [][]byte, _ [][]*x509.Certificate) error {
+		var certs []*x509.Certificate
+		for _, rawCert := range rawCerts {
+			cert, err := x509.ParseCertificate(rawCert)
+			if err != nil {
+				return err
+			}
+			certs = append(certs, cert)
+		}
+
+		roots, err := p.GetRoots()
+		if err != nil {
+			return err
+		}
+		if _, err := VerifyPeerCertificate(certs, roots, expectPeer); err != nil {
+			return err
+		}
+		return nil
+	}
+}
+
+func (p *TLSPeer) DialGRPC(ctx context.Context, addr string, expectPeer ExpectPeerFunc, opts ...grpc.DialOption) (*grpc.ClientConn, error) {
+	config, err := p.GetConfig(ctx, expectPeer)
+	if err != nil {
+		return nil, err
+	}
+	return grpc.DialContext(ctx, addr, append([]grpc.DialOption{
+		grpc.WithTransportCredentials(credentials.NewTLS(config)),
+	}, opts...)...)
```

Comment:
Do we want to inject our logger here to expose handshake errors we wouldn't get otherwise?

Reply:
Hmm, are you talking about adapting to the grpclog interface or wrapping the transport credentials with something that logs errors?

Reply:
I was just thinking about how we need to pass custom dialer in SPIRE and perform TLS handshake directly, otherwise we get crappy TLS errors from gRPC

Reply:
Hmm. I will see what I can do...

Reply:
Client-side handshake errors are logged pretty verbosely so I just wired up the peer logger to the VerifyPeerCertificate wrapper and logged errors from there.